wren-ai-service/src/pipelines/generation/data_assistance.py

```diff
@@ -0,0 +1,198 @@
+import asyncio
+import logging
+import sys
+from pathlib import Path
+from typing import Any, Optional
+
+import orjson
+from hamilton import base
+from hamilton.experimental.h_async import AsyncDriver
+from haystack.components.builders.prompt_builder import PromptBuilder
+from langfuse.decorators import observe
+from pydantic import BaseModel
+
+from src.core.pipeline import BasicPipeline
+from src.core.provider import LLMProvider
+from src.utils import async_timer, timer
+
+logger = logging.getLogger("wren-ai-service")
+
+
+data_assistance_system_prompt = """
+### TASK ###
+You are a data analyst great at answering user's questions about given database schema.
+Please carefully read user's question and database schema to answer it in easy to understand manner
+using the Markdown format. Your goal is to help guide user understand its database!
+
+### INSTRUCTIONS ###
+
+- Answer must be in the same language user specified.
+- There should be proper line breaks, whitespace, and Markdown formatting(headers, lists, tables, etc.) in your response.
+- If the language is Traditional/Simplified Chinese, Korean, or Japanese, the maximum response length is 150 words; otherwise, the maximum response length is 110 words.
+
+### OUTPUT FORMAT ###
+Please provide your response in proper Markdown format.
+"""
+
+data_assistance_user_prompt_template = """
+### DATABASE SCHEMA ###
+{% for db_schema in db_schemas %}
+    {{ db_schema }}
+{% endfor %}
+
+### INPUT ###
+User's question: {{query}}
+Language: {{language}}
+
+Please think step by step
+"""
+
+
+## Start of Pipeline
+@timer
+@observe(capture_input=False)
+def prompt(
+    query: str,
+    db_schemas: list[str],
+    language: str,
+    prompt_builder: PromptBuilder,
+) -> dict:
+    logger.debug(f"query: {query}")
+    logger.debug(f"db_schemas: {db_schemas}")
+    logger.debug(f"language: {language}")
+
+    return prompt_builder.run(query=query, db_schemas=db_schemas, language=language)
+
+
+@async_timer
+@observe(as_type="generation", capture_input=False)
+async def data_assistance(prompt: dict, generator: Any, query_id: str) -> dict:
+    logger.debug(f"prompt: {orjson.dumps(prompt, option=orjson.OPT_INDENT_2).decode()}")
+
+    return await generator.run(prompt=prompt.get("prompt"), query_id=query_id)
+
+
+## End of Pipeline
+
+
+class DataAssistanceResult(BaseModel):
+    results: str
+
+
+DATA_ASSISTANCE_MODEL_KWARGS = {"response_format": {"type": "text"}}
+
+
+class DataAssistance(BasicPipeline):
+    def __init__(
+        self,
+        llm_provider: LLMProvider,
+        **kwargs,
+    ):
+        self._user_queues = {}
```

Comment:
I think we could consider using a TTL cache to store it, which would help prevent memory usage from increasing over time.

Reply:
another option is to store them in the cache of ask service, but we need to further consider the design.